File: bff/src/main/java/jp/co/ogis_ri/nautible/app/payment/bff/outbound/rest/PaymentCashRepository.java
package jp.co.ogis_ri.nautible.app.payment.bff.outbound.rest;

import javax.ws.rs.DELETE;
import javax.ws.rs.POST;
import javax.ws.rs.PUT;
import javax.ws.rs.Path;

import org.eclipse.microprofile.rest.client.inject.RegisterRestClient;
import org.jboss.resteasy.annotations.jaxrs.PathParam;

import jp.co.ogis_ri.nautible.app.payment.bff.domain.Payment;

/**
 * 決済処理をバックエンド（代引き）とやり取りするインターフェース
 */
@Path("/nautible-app-payment-cash")
@RegisterRestClient
public interface PaymentCashRepository {

    @POST
    @Path("/method/payment/")
    Payment create(Payment payment);

    @DELETE
    @Path("/method/payment/{paymentNo}")
    Payment delete(@PathParam String paymentNo);

    @PUT
    @Path("/methodpayment/")
    Payment update(Payment payment);

    //Payment getByRequestId(String requestId);
}

File: bff/src/outbound/payment_cash_repository.go
package outbound

import (
	"bytes"
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	outbound "payment-bff/generate/client/payment"
	domain "payment-bff/src/domain"
)

type PaymentCashRepository struct{}

func NewPaymentCashRepository() domain.CashRepository {
	paymentCashRepository := PaymentCashRepository{}
	return &paymentCashRepository
}

// 決済登録を行うリポジトリ
func (p *PaymentCashRepository) CreatePayment(request *domain.PaymentItem) (*domain.PaymentItem, error) {
	c, err := outbound.NewClient("http://localhost:3500/v1.0/invoke/nautible-app-ms-payment-cash/method")
	if err != nil {
		panic(err)
	}
	var restCreatePayment outbound.RestCreatePayment
	restCreatePayment.CustomerId = request.CustomerId
	restCreatePayment.OrderDate = request.OrderDate
	restCreatePayment.OrderNo = request.OrderNo
	restCreatePayment.PaymentType = request.PaymentType
	restCreatePayment.RequestId = request.RequestId
	restCreatePayment.TotalPrice = request.TotalPrice
	requestJson, err := json.Marshal(restCreatePayment)
	if err != nil {
		fmt.Println(err)
		return nil, err
	}
	buf := bytes.NewBuffer(requestJson)

	res, err := c.CreateWithBody(context.Background(), "application/json", buf)
	defer res.Body.Close()
	if err != nil {
		return nil, err
	}
	if res.StatusCode == 200 || res.StatusCode == 201 {
		buf := new(bytes.Buffer)
		io.Copy(buf, res.Body)
		var result domain.PaymentItem
		json.Unmarshal(buf.Bytes(), &result)
		return &result, nil
	}
	return nil, errors.New("http response not success")
}

// 決済データ取得を行うリポジトリ
func (p *PaymentCashRepository) GetByPaymentNo(paymentNo string) (*domain.PaymentItem, error) {
	fmt.Println("Rest GetByPaymentNo")
	c, err := outbound.NewClientWithResponses("http://localhost:3500/v1.0/invoke/nautible-app-ms-payment-cash/method")
	if err != nil {
		return &domain.PaymentItem{}, err
	}

	// http.Response として返却
	res, err := c.GetByPaymentNoWithResponse(context.Background(), paymentNo)
	if err != nil {
		return &domain.PaymentItem{}, err
	}
	var model domain.PaymentItem
	json.NewDecoder(bytes.NewReader(res.Body)).Decode(&model)
	return &model, err
}
